File: src/types/Service.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const Service = DB.Entity(import.meta.url, {
  name: "Service",
  namePlural: "Services",
  type: () =>
    DB.Object({
      availability: DB.Required({
        comment: "Defines for which creature type(s) the service is available.",
        type: DB.Array(DB.IncludeIdentifier(ServiceAvailability), {
          minItems: 1,
          uniqueItems: true,
        }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "Service",
        DB.Object({
          name: DB.Required({
            comment: "The service’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          description: DB.Required({
            comment: "The service’s description.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const ServiceAvailability = DB.Enum(import.meta.url, {
  name: "ServiceAvailability",
  values: () => ({
    SummonedCreatures: DB.EnumCase({ type: null }),
    Monstrosities: DB.EnumCase({ type: null }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/Service.swift
import FileDB

@Model
public struct Service {

      /// Defines for which creature type(s) the service is available.
      @MinItems(1)
      @UniqueItems
      let availability: [ServiceAvailability]

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // ServiceTranslation

        /// The service’s name.
        @MinLength(1)
        let name: String

        /// The service’s description.
        @MinLength(1)
        @Markdown
        let description: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

@ModelEnum
public enum ServiceAvailability {
    case SummonedCreatures
    case Monstrosities
}
